### PR TITLE
Make some types and methods related to Polonius + Miri public

### PR DESCRIPTION
**Pull request by willcrichton**
We have a tool, [Aquascope](https://github.com/cognitive-engineering-lab/aquascope/), which uses Polonius and Miri to visualize the compile-time and run-time semantics of a Rust program. Changes in the last few months to both APIs have hidden away details we depend upon. This PR re-exposes some of those details, specifically:

**Polonius:**
- `BorrowSet` and `BorrowData` are added to `rustc_borrowck::consumers`, and their fields are made `pub` instead of `pub(crate)`. We need this to interpret the `BorrowIndex`es generated by Polonius.
- `BorrowSet::build` is now `pub`. We need this because the borrowck API doesn't provide access to the `BorrowSet` constructed during checking.
- `PoloniusRegionVid` is added to `rustc_borrowck::consumers`. We need this because it's also contained in the Polonius facts.

**Miri:**
- `InterpCx::local_to_op` is now a special case of `local_at_frame_to_op`, which allows querying locals in any frame. We need this because we walk the whole stack at each step to collect the state of memory.
- `InterpCx::layout_of_local` is now `pub`. We need this because we need to know the layout of every local at each step.

If these changes go against some design goal for keeping certain types private, please let me know so we can hash out a better solution. Additionally, if there's a better way to document that it's important that certain types stay public, also let me know. For example, `BorrowSet` was previously public but was hidden in 6676cec, breaking our build.

cc @RalfJung @nnethercote @gavinleroy

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -706,17 +706,27 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
...
+
+    /// Read from a local of a given frame.
     /// Will not access memory, instead an indirect `Operand` is returned.
     ///
-    /// This is public because it is used by [priroda](https://github.com/oli-obk/priroda) to get an
+    /// This is public because it is used by [priroda](https://github.com/oli-obk/priroda) and
+    /// [Aquascope](https://github.com/cognitive-engineering-lab/aquascope/) to get an
     /// OpTy from a local.
-    pub fn local_to_op(
+    pub fn local_at_frame_to_op(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:724]: This calls `after_local_read` which has no way of knowing that the local is from a different frame... so things might go wrong in Miri here.

Please extend `after_local_read` to also receive the `&Frame`.

File: compiler/rustc_borrowck/src/borrow_set.rs [hunk 2]
```diff
@@ -55,17 +55,17 @@ pub(crate) enum TwoPhaseActivation {
...
+    pub kind: mir::BorrowKind,
     /// The region for which this borrow is live
-    pub(crate) region: RegionVid,
+    pub region: RegionVid,
     /// Place from which we are borrowing
-    pub(crate) borrowed_place: mir::Place<'tcx>,
+    pub borrowed_place: mir::Place<'tcx>,
     /// Place to which the borrow was stored
-    pub(crate) assigned_place: mir::Place<'tcx>,
+    pub assigned_place: mir::Place<'tcx>,
```
Comment (RalfJung) [inline compiler/rustc_borrowck/src/borrow_set.rs:68]: Do you really need read and write access to all these fields?

If read access suffices, maybe add public getters instead?

**Comment by lqd** [comment]
> Additionally, if there's a better way to document that it's important that certain types stay public, also let me know. For example, `BorrowSet` was previously public

Adding a comment to each def site of interest will help: the fact that they are public is not enough, neither is that they're referenced from `consumers`. AFAICT this code is has no stability guarantee, so thigns can (and will) still break but a comment explaining why things are public can help lower the chances of accidental breakage. This is already done in multiple places.

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 3]
```diff
@@ -706,17 +706,27 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
...
+        layout: Option<TyAndLayout<'tcx>>,
+    ) -> InterpResult<'tcx, OpTy<'tcx, M::Provenance>> {
+        self.local_at_frame_to_op(self.frame(), local, layout)
+    }
+
+    /// Read from a local of a given frame.
     /// Will not access memory, instead an indirect `Operand` is returned.
     ///
-    /// This is public because it is used by [priroda](https://github.com/oli-obk/priroda) to get an
+    /// This is public because it is used by [priroda](https://github.com/oli-obk/priroda) and
```
Comment (lqd) [inline compiler/rustc_const_eval/src/interpret/operand.rs:721]: IIRC priroda doesn't work anymore (right @oli-obk?), if so, maybe remove the outdated mention here? 

(priroda could be a good use-case for smir and its project goal though)

File: compiler/rustc_const_eval/src/interpret/operand.rs (hunk 1 above)
Comment (willcrichton) [reply compiler/rustc_const_eval/src/interpret/operand.rs:723]: Done.

File: compiler/rustc_borrowck/src/borrow_set.rs (hunk 2 above)
Comment (willcrichton) [reply compiler/rustc_borrowck/src/borrow_set.rs]: I don't need write access. I restored the `pub(crate)` visibility and added getters for `BorrowSet` and `BorrowData`.

**Comment by RalfJung** [comment]
r=me on the interpret and Miri parts.
It's a bit odd that reading locals now works for all frames but writing only works for the current frame... but maybe that's not bad, it's generally not a thing to directly write the local of another frame.

**Comment by lqd** [comment]
r=me on the borrowck/polonius parts

**Comment by RalfJung** [comment]
I think that's all the parts.

@bors r=RalfJung,lqd rollup

**Pull request merged**